qos-service-api/src/main/java/com/palantir/atlasdb/qos/QosClient.java

```diff
@@ -16,6 +16,25 @@
 
 package com.palantir.atlasdb.qos;
 
+import java.util.function.Function;
+import java.util.function.Supplier;
+
 public interface QosClient {
-    void checkLimit();
+
+    interface ReadQuery<T, E extends Exception> {
+        T execute() throws E;
+    }
+
+    interface WriteQuery<E extends Exception> {
+        void execute() throws E;
+    }
+
+    <T, E extends Exception> T executeRead(
+            Supplier<Integer> estimatedWeight,
+            ReadQuery<T, E> query,
+            Function<T, Integer> weigher) throws E;
```

Comment:
`weigher` seems sub-optimally named - doesn't capture that this gives us the actual weight, after query execution. `actualWeight` (or clunkier, `actualWeightGivenResult`) would dovetail nicely with `estimatedWeight`.

Reply:
yeah, this is addressed somewhat in the refactor at https://github.com/palantir/atlasdb/pull/2697